packages/server/src/course.rs

```diff
@@ -234,21 +239,63 @@ pub fn parse_copy_paste_data(data: &str) -> Result<Vec<CourseStatus>, Error> {
             sport_courses.push(course_status);
             continue;
         }
-        *courses
-            .entry(course_status.course.id.clone())
-            .or_insert(course_status) = course_status.clone();
+        if line.contains('*') {
+            // If a student decides to retake a course for which he already had a grade,
+            // and then ends up not receiving a grade (לא השלים) for that course,
+            // The previous grade for this course is the valid one.
+            // Nevertheless, the previous grade will appear with an asterisk (*) in the grades pdf.
+            // Thus, to make sure we don't ignore these cases, we have to save a list of every asterisk-marked course,
+            // and then search this list for courses who fall in this particular case, and fix their grade.
+            asterisk_courses.push(course_status);
+        } else {
+            *courses
+                .entry(course_status.course.id.clone())
+                .or_insert(course_status) = course_status.clone();
+        }
     }
     let mut vec_courses: Vec<_> = courses.into_values().collect();
+
+    // Fix the grades for said courses
+    set_grades_for_uncompleted_courses(&mut vec_courses, asterisk_courses.clone());
+
     vec_courses.append(&mut sport_courses);
+
     if vec_courses.is_empty() {
         return Err(ErrorBadRequest("Bad Format"));
     }
     Ok(vec_courses)
 }
 
-fn parse_course_status_pdf_format(line: String) -> Result<(Course, Option<Grade>), Error> {
+fn set_grades_for_uncompleted_courses(
+    courses: &mut Vec<CourseStatus>,
+    asterisk_courses: Vec<CourseStatus>,
+) {
+    // The canditate course statuses are those with uncomplete (לא השלים) grades.
+    // For each uncompleted course status, we iterate the asterisk list in reverse to find
+    // the closest (most chronologically advanced) course status with a grade (anything other than NotComplete (לא השלים)).
+    // This course status will replace the old one.
+    let uncompleted_courses = courses
+        .iter_mut()
+        .filter(|c| c.grade == Some(Grade::NotComplete))
+        .collect::<Vec<_>>();
```

Comment:
let's keep the convention consistent between here and line 256.
I mean whether write the "Vec<_>" near to the declaration of "uncompleted_courses" as in line 256.
Or in line 256write the "Vec<_>" in the collect method as you did here.
I am fine with either choice.

Reply:
fixed